pecos/core/utils/mmap_util.hpp

```diff
@@ -0,0 +1,530 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+#ifndef __MMAP_UTIL_H__
+#define __MMAP_UTIL_H__
+
+/* Memory-mapped module version.
+ * Must be incremented by 1 everytime when non-backward-compatible change is made. */
+#define __MMAP_UTIL_VER 1
+
+
+#include <cstring>
+#include <fcntl.h>
+#include <memory>
+#include <stdexcept>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <vector>
+#include <string>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include "file_util.hpp"
+
+
+namespace pecos {
+
+namespace mmap_util {
+
+namespace details_ { // namespace for Module Private classes
+
+/* Type constraint for mmap save/load.
+ * Please beware, this is only a minimum requirement and not a guarantee that the type is platform-portable for mmap.
+ * If the class/struct members need padding, please double-check mmap file portability on your own. */
+template<class Type>
+constexpr bool IsPlainOldData = std::is_trivially_copyable<Type>::value && std::is_standard_layout<Type>::value;
+
+template<typename Type, typename Ret=bool>
+using if_simple_serializable = std::enable_if_t<IsPlainOldData<Type>, Ret>;
+
+
+/*
+ * Module Private Class to store signature and metadata of all data blocks in mainbody for a memory-mapped file.
+ */
+class MmapMetadata_ {
```

Comment:
I would suggest to have a trailing "_" for private members (variables, or methods) but not for the class definition given that all the classes are already wrapped in a privately hinted details_ namespace.